pkg/util/http.go

```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
 	return apiKeyRegExp.ReplaceAllString(message, apiKeyReplacement)
 }
+
+// HTTPHeaders returns a http headers including various basic information (User-Agent, Content-Type...).
+func HTTPHeaders() map[string]string {
+	av, _ := version.New(version.AgentVersion, version.Commit)
+	return map[string]string{
+		"User-Agent":   fmt.Sprintf("Datadog Agent/%s", av.GetNumber()),
+		"Content-Type": "application/x-www-form-urlencoded",
+		"Accept":       "text/html, */*",
+	}
+}
+
+// GetProxyTransportFunc return a proxy function for a http.Transport that
+// would return the right proxy depending on the configuration.
+func GetProxyTransportFunc(p *config.Proxy) func(*http.Request) (*url.URL, error) {
+	return func(r *http.Request) (*url.URL, error) {
+		// check no_proxy list first
+		if matches, matchingNoProxy := useProxy(canonicalAddr(r.URL), p.NoProxy); !matches {
+			log.Debugf("'%s' matches no_proxy list item '%s': not using any proxy", SanitizeURL(r.URL.String()), matchingNoProxy)
+			return nil, nil
+		}
+
+		// check proxy by scheme
+		confProxy := ""
+		if r.URL.Scheme == "http" {
+			confProxy = p.HTTP
+		} else if r.URL.Scheme == "https" {
+			confProxy = p.HTTPS
+		} else {
+			log.Warnf("Proxy configuration does not support scheme '%s'", r.URL.Scheme)
+		}
+
+		if confProxy != "" {
+			proxyURL, err := url.Parse(confProxy)
+			if err != nil {
+				err := fmt.Errorf("Could not parse the proxy URL for scheme %s from configuration: %s", r.URL.Scheme, err)
+				log.Error(err.Error())
+				return nil, err
+			}
+			userInfo := ""
+			if proxyURL.User != nil {
+				if _, isSet := proxyURL.User.Password(); isSet {
+					userInfo = "*****:*****@"
+				} else {
+					userInfo = "*****@"
+				}
+			}
+
+			log.Debugf("Using proxy %s://%s%s for URL '%s'", proxyURL.Scheme, userInfo, proxyURL.Host, SanitizeURL(r.URL.String()))
+			return proxyURL, nil
+		}
+
+		// no proxy set for this request
+		return nil, nil
+	}
+}
+
+// CreateHTTPTransport creates an *http.Transport for use in the agent
+func CreateHTTPTransport() *http.Transport {
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: config.Datadog.GetBool("skip_ssl_validation"),
+	}
+
+	if config.Datadog.GetBool("force_tls_12") {
+		tlsConfig.MinVersion = tls.VersionTLS12
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	if proxies := config.GetProxies(); proxies != nil {
+		transport.Proxy = GetProxyTransportFunc(proxies)
+	}
+	return transport
+}
+
+// useProxy reports whether requests to addr should use a proxy,
+// according to the noProxy entries, and if a proxy should be used it
+// returns the matching noProxy entry.
+// addr is always a canonicalAddr with a host and port.
+// Copied from `net/http/transport.go`, modified to use the passed noProxy values
+// instead of pulling directly from the no_proxy env var, and to return the
+// matched noProxy entry.
+func useProxy(addr string, noProxy []string) (matches bool, matchingNoProxy string) {
```

Comment:
When setting the environment variable `HTTP_PROXY` all call to the cloud providers metadata endpoint are going through the proxy.

I think this function should white-list the those IP:
- `169.254.169.254` for ec2, gce and azure
- `169.254.170.2` for ecs (maybe not this one ?)